### PR TITLE
Sparql data provider new options

### PR DESCRIPTION
**Pull request by yuricus**
Extending sparql data provider to support new options. Status right now it WIP, I will update this with abstract graph support soon.

### CODE REVIEW

File: src/ontodia/data/sparql/sparqlDataProvider.ts [hunk 1]
```diff
@@ -268,13 +276,49 @@ export class SparqlDataProvider implements DataProvider {
...
     }
 
     executeSparqlConstruct(query: string) : Promise<Triple[]> {
         const method = this.options.queryMethod ? this.options.queryMethod : SparqlQueryMethod.GET;
-        return executeSparqlConstruct(this.options.endpointUrl, query, method);
+        return executeSparqlConstruct(this.options.endpointUrl, query, method, this.options.queryFunction);
     }
+
+    createRefQueryPart(params: { elementId: string; linkId?: string; direction?: 'in' | 'out'}) {
+    let refQueryPart = '';
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProvider.ts]: Something is wrong with the indentation here.

File: src/ontodia/data/sparql/sparqlDataProvider.ts [hunk 2]
```diff
@@ -22,6 +22,13 @@ import { SparqlDataProviderSettings, OWLStatsSettings } from './sparqlDataProvid
 
 export enum SparqlQueryMethod { GET = 1, POST }
 
+export type QueryFunction = (params: {
+    url: string,
+    body: string,
+    headers: any,
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProvider.ts]: Maybe `headers` parameter should have type `{ [header: string]: string }` instead?

File: src/ontodia/data/sparql/sparqlDataProvider.ts [hunk 3]
```diff
@@ -313,16 +357,20 @@ export function executeSparqlQuery<Binding>(endpoint: string, query: string, met
             return response.json();
         } else {
             const error = new Error(response.statusText);
-            (<any>error).response = response;
+            (<any> error).response = response;
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProvider.ts]: Why add a space here?

File: src/ontodia/data/sparql/sparqlDataProviderSettings.ts [hunk 4]
```diff
@@ -1,17 +1,18 @@
 /**
  * this is dataset-schema specific settings
  */
-export interface SparqlDataProviderSettings {
+    export interface SparqlDataProviderSettings {
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProviderSettings.ts]: Broken indentation?

File: src/ontodia/data/sparql/sparqlDataProviderSettings.ts [hunk 5]
```diff
@@ -93,7 +107,44 @@ export interface FullTextSearchSettings {
...
+    refElementQueryIn: '?inst ?link \${refElementIRI} .',
+
+    linksInfoQuery: `SELECT ?source ?type ?target
+            WHERE {
+                ?source ?type ?target.
+                VALUES (?source) {\${ids}}
+                VALUES (?target) {\${ids}}                
+            }`,
+
+    defaultPrefix: ``,
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProviderSettings.ts]: Don't use template literals when it's just a single-line string with no template substitutions.

File: src/ontodia/data/sparql/sparqlDataProviderSettings.ts [hunk 6]
```diff
@@ -67,6 +73,14 @@ export interface SparqlDataProviderSettings {
      * imposes additional filtering on elements within filter
      */
     filterAdditionalRestriction: string;
+
+    /**
+     * Link reference resolving. Developer can use it to transform how link navigation is performed.
+     */
+    refElementLinkQueryOut: string; // '\${refElementIRI} \${refElementLinkIRI} ?inst . FILTER ISIRI(?inst)',
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProviderSettings.ts]: You should move the default values from inline comments to JSDoc `@default` delcaration.

File: src/ontodia/data/sparql/sparqlDataProvider.ts [hunk 7]
```diff
@@ -313,16 +357,20 @@ export function executeSparqlQuery<Binding>(endpoint: string, query: string, met
...
         }
     });
 };
 
-export function executeSparqlConstruct(endpoint: string, query: string, method: SparqlQueryMethod): Promise<Triple[]> {
+export function executeSparqlConstruct(
+    endpoint: string,
+    query: string,
+    method: SparqlQueryMethod,
+    queryFunction: QueryFunction): Promise<Triple[]> {
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProvider.ts]: Move closing parenthesis to the next line (there should be a linting warning but I'm not sure that it detects this right now).

File: src/ontodia/data/sparql/sparqlDataProviderSettings.ts [hunk 8]
```diff
@@ -1,17 +1,18 @@
...
      */
     defaultPrefix: string;
+
     /**
      *  property to use as label in schema (classes, properties)
      */
     schemaLabelProperty: string;
 
-    /**
+     /**
```
Comment (AlexeyMz) [inline src/ontodia/data/sparql/sparqlDataProviderSettings.ts]: Misaligned JSDoc indentation here (extra space before `/**`).

File: src/ontodia/data/sparql/sparqlDataProvider.ts (hunk 2 above)
Comment (yuricus) [inline src/ontodia/data/sparql/sparqlDataProvider.ts]: Actually it's partial fetch parameters, I would prefer them to be finite and described.

**Pull request merged**